Rpc/Serdes.cs

```diff
@@ -0,0 +1,97 @@
+
+    public async ValueTask<IDisposable> LockAsync(CancellationToken ct = default)
+    {
+        await _semaphore.WaitAsync(ct);
+        return new Lock(_semaphore);
+    }
+
+    private class Lock(SemaphoreSlim semaphore) : IDisposable
+    {
+        public void Dispose()
+        {
+            semaphore.Release();
+            GC.SuppressFinalize(this);
+        }
+    }
+}
+
+/// <summary>
+///     Serdes provides serialization and deserialization of messages read from a Stream.
+/// </summary>
+public class Serdes<TS, TR>
+    where TS : RpcMessage<TS>, IMessage<TS>
+    where TR : RpcMessage<TR>, IMessage<TR>, new()
+{
+    private const int MaxMessageSize = 0x1000000; // 16MiB
+
+    private readonly MessageParser<TR> _parser = new(() => new TR());
+
+    private readonly RaiiSemaphoreSlim _readLock = new(1, 1);
+    private readonly RaiiSemaphoreSlim _writeLock = new(1, 1);
+
+    /// <summary>
+    ///     Encodes and writes a message to the Stream. Only one message will be written at a time.
+    /// </summary>
+    /// <param name="conn">Stream to write the encoded message to</param>
+    /// <param name="message">Message to encode and write</param>
+    /// <param name="ct">Optional cancellation token</param>
+    /// <exception cref="ArgumentException">If the message exceeds the maximum message size of 16 MiB</exception>
+    public async Task WriteMessage(Stream conn, TS message, CancellationToken ct = default)
+    {
+        using var _ = await _writeLock.LockAsync(ct);
+
+        var mb = message.ToByteArray();
+        if (mb.Length > MaxMessageSize)
+            throw new ArgumentException($"Marshalled message size {mb.Length} exceeds maximum {MaxMessageSize}");
+
+        var lenBytes = new byte[4];
+        BinaryPrimitives.WriteUInt32BigEndian(lenBytes, (uint)mb.Length);
+        await conn.WriteAsync(lenBytes, ct);
+        await conn.WriteAsync(mb, ct);
+    }
+
+    /// <summary>
+    ///     Reads a decodes a single message from the stream. Only one message will be read at a time.
+    /// </summary>
+    /// <param name="conn">Stream to read the message from</param>
+    /// <param name="ct">Optional cancellation token</param>
+    /// <returns>Decoded message</returns>
+    /// <exception cref="IOException">Could not decode the message</exception>
+    public async Task<TR> ReadMessage(Stream conn, CancellationToken ct = default)
+    {
+        using var _ = await _readLock.LockAsync(ct);
+
+        var lenBytes = new byte[sizeof(uint)];
+        await conn.ReadExactlyAsync(lenBytes, ct);
+        var len = BinaryPrimitives.ReadUInt32BigEndian(lenBytes);
+        if (len > MaxMessageSize)
+            throw new IOException($"Received message size {len} exceeds maximum {MaxMessageSize}");
+
+        var msgBytes = new byte[len];
+        await conn.ReadExactlyAsync(msgBytes, ct);
+
+        try
+        {
+            var msg = _parser.ParseFrom(msgBytes);
+            if (msg?.RpcField is null)
```

Comment:
Messages from the tunnel that aren't a response and don't require a response have a null `RPC` field. My understanding of the design is that the presence of `MsgID` indicates whether a response is expected.

Reply:
@spikecurtis thoughts on changing this behavior in the protocol? I think all messages should have RPC set and have a message ID. If you want to enforce on each end that replies can only be sent to messages that expect one, we could add a `ExpectReply bool` to RPC or something perhaps

Reply:
It's like having a database table where the id field is optional... if not all messages have an ID then it should be called something else, like `reply_with_id`

During debugging it might be helpful to also log sent and received messages, and it's nice if every message has an ID that can be logged on both ends of the connection

Reply:
I've updated the behavior to match Go for now and I talked with Spike today about making some changes to the name of the MsgId field which I'll do later